### PR TITLE
Persist execution context in storage target

### PR DESCRIPTION
**Pull request by cisaacstern**
This is a WIP to address @rabernat's suggestion in https://github.com/pangeo-forge/pangeo-forge-recipes/issues/37#issuecomment-1118899827 that we persist execution context information in the storage target, which is the first part of a solution for #37 and https://github.com/pangeo-forge/user-stories/issues/5.

Here's a rough idea of how this might look:

- `BaseRecipe.get_execution_context` compiles a dict of metadata we want to capture
- Each recipe class individually elects when/where this will be persisted to the target. The stage in which this occurs will be specific to each recipe class (`finalize_target` for XarrayZarrRecipe, a different stage for HDFReferenceRecipe, etc.). As will the exact location in the target (`.zmetadata` for  XarrayZarrRecipe, somewhere else for other storage formats, etc.).

### CODE REVIEW

File: pangeo_forge_recipes/recipes/base.py [hunk 1]
```diff
@@ -50,6 +52,14 @@ def to_beam(self):
     def sha256(self):
         return dataclass_sha256(self, ignore_keys=self._hash_exclude_)
 
+    def get_execution_context(self):
+        return dict(
+            # See https://stackoverflow.com/a/2073599 re: version
+            version=pkg_resources.require("registrar")[0].version,
```
Comment (rabernat) [inline pangeo_forge_recipes/recipes/base.py:58]: Why track the registrar version? Ideally the target contents don't depend on the registrar. I was thinking we could track the pangeo_forge_recipes package version.

Comment (cisaacstern) [reply pangeo_forge_recipes/recipes/base.py:58]: Good catch! I copy-and-pasted this from the registrar, and forgot to update that.

Also this comment reminds me to add this PR to my TODO list of the week. 😄

Comment (cisaacstern) [reply pangeo_forge_recipes/recipes/base.py]: fixed in https://github.com/pangeo-forge/pangeo-forge-recipes/pull/359/commits/3ca6066b4f9ad6b19747f70a0c5ae11b779589da

**Comment by cisaacstern** [comment]
The latest test failures are from `test_XarrayZarrRecipe.py::test_process`, and stem from the fact that in #349 we punted on how to hash preprocessor functions. In this PR, we're calling `recipe.get_execution_context()` (which includes a `recipe.sha256()` call) on every execution, so the tests which include preprocessors fail. The options would seem to be:

1. Find a way to deterministically hash Python functions. (I've previously wondered if the function objects `__code__` attribute might offer some path forward here...)
2. Don't set a `"pangeo-forge:recipe_hash"` if the instance contains a preprocessor.
3. Do set a `"pangeo-forge:recipe_hash"` for recipes with a preprocessor, but exclude the preprocessor from the calculation. We might think of this as thematically similar to the `unsafe_hash` option for built-in dataclasses.

Preprocessors are a very useful feature, as (will be) hashes, so my preference would be to try to find a way to achieve option 1 if possible. I will take a brief pause on this PR to look into this, but if this seems infeasible, as a backup plan we could move forward with option 2 (not setting hashes for recipes with preprocessors). It seems to me that option 3 could lead to some very confusing situations which we'd prefer to avoid.

File: tests/recipe_tests/test_XarrayZarrRecipe.py [hunk 2]
```diff
@@ -120,7 +131,7 @@ def test_recipe(recipe_fixture, execute_recipe):
     rec = RecipeClass(file_pattern, **kwargs)
     execute_recipe(rec)
     ds_actual = xr.open_zarr(target.get_mapper()).load()
-    xr.testing.assert_identical(ds_actual, ds_expected)
+    xr.testing.assert_identical(drop_execution_context_attrs(ds_actual), ds_expected)
```
Comment (rabernat) [inline tests/recipe_tests/test_XarrayZarrRecipe.py:134]: I'm not a fan of having to include this statement in every test.

Instead, let's write a wrapper around `xr.testing.assert_identical` like this

```python
def assert_identical(ds1, ds2):
    xr.testing.assert_identical(
        drop_execution_context_attrs(ds1),
        drop_execution_context_attrs(ds2)
    )
```

and call it from the individual tests.

File: docs/pangeo_forge_recipes/recipe_user_guide/execution.md [hunk 3]
```diff
@@ -91,3 +91,26 @@ with beam.Pipeline() as p:
...
+    "pangeo-forge:recipe_hash": "{recipe hash as returned by `recipe.sha256()`}"
+    "pangeo-forge:inputs_hash": "{file pattern hash as returned by `recipe.file_pattern.sha256()`}"
+}
+```
+
+Each recipe class defines where to store this metadata:
+
+- `XarrayZarrRecipe`: Added to Zarr group attributes, and therefore also available via the
+`xarray.Dataset.attrs` when opening Zarr stores with xarray.
+- `HDFReferenceRecipe`:
```
Comment (rabernat) [inline docs/pangeo_forge_recipes/recipe_user_guide/execution.md:113]: It should be straightforward to add the metadata to the kerchunk dataset as well. But we can leave that for a future PR.

**Review by rabernat — COMMENTED** [review]
This is looking great! Just one nit pick about the testing syntax.

File: docs/pangeo_forge_recipes/recipe_user_guide/execution.md (hunk 3 above)
Comment (cisaacstern) [reply docs/pangeo_forge_recipes/recipe_user_guide/execution.md]: 👍 [62183f3](https://github.com/pangeo-forge/pangeo-forge-recipes/pull/359/commits/62183f35badc963d36b0622954d03d76865a494f)

File: tests/recipe_tests/test_XarrayZarrRecipe.py (hunk 2 above)
Comment (cisaacstern) [reply tests/recipe_tests/test_XarrayZarrRecipe.py]: 👍 b781e30

**Comment by cisaacstern** [comment]
Thanks for the review, Ryan. `inspect.getsource` is very cool, hadn't seen it before! I think that's the last thing we need here, will add that now.

**Review by rabernat — COMMENTED** [review]
Sorry I never merged this! I just see one small issue, then LGTM.

File: tests/recipe_tests/test_XarrayZarrRecipe.py [hunk 4]
```diff
@@ -21,6 +21,26 @@
 from pangeo_forge_recipes.storage import MetadataTarget, StorageConfig
 
 
+def drop_execution_context_attrs(ds: xr.Dataset) -> xr.Dataset:
+    """Drop pangeo-forge execution context attrs from a dataset."""
+
+    ds_copy = ds.copy(deep=True)
```
[the remark this answers is not in the document]
Comment (cisaacstern) [reply tests/recipe_tests/test_XarrayZarrRecipe.py:27]: I'm not sure! It's "just" the tests so happy to use regular copy and see how it goes.

TBH, I've now run into enough weird behavior of copied dictionaries that are _not_ deep copies (due to shared references with the original, I guess), that I just reflexively make all dictionary copies deep now. 😅

Comment (cisaacstern) [inline tests/recipe_tests/test_XarrayZarrRecipe.py:27]:
```suggestion
    ds_copy = ds.copy()
```

File: pangeo_forge_recipes/recipes/base.py (hunk 1 above)
Comment (KoscheyAV1) [reply pangeo_forge_recipes/recipes/base.py]: []()

**Pull request merged**